spring-cloud-azure-keyvault-config/src/main/java/com/microsoft/azure/spring/cloud/keyvault/config/KeyVaultPropertySource.java

```diff
@@ -6,24 +6,66 @@
 
 package com.microsoft.azure.spring.cloud.keyvault.config;
 
+import com.microsoft.azure.PagedList;
+import com.microsoft.azure.keyvault.KeyVaultClient;
+import com.microsoft.azure.keyvault.models.SecretItem;
 import org.springframework.core.env.EnumerablePropertySource;
 
+import java.util.concurrent.ConcurrentHashMap;
+
 /**
  * Retrieve all properties from the Azure Key Vault instance.
  */
-// TODO: Use Key Vault source class here.
-public class KeyVaultPropertySource extends EnumerablePropertySource<Object> {
-    public KeyVaultPropertySource(String name, Object source) {
-        super(name, source);
+public class KeyVaultPropertySource extends EnumerablePropertySource<KeyVaultClient> {
+
+    private final KeyVaultClient keyVaultClient;
+    private final String vaultUrl;
+
+    // Saves names of all secrets in a Key Vault. Key is secret name. Value is null.
+    private final ConcurrentHashMap<String, String> secretNames = new ConcurrentHashMap<>();
+    // Saves values of each secret. Key is secret name. Value is secret value.
+    private final ConcurrentHashMap<String, String> secretValues = new ConcurrentHashMap<>();
+
+    public KeyVaultPropertySource(String name, KeyVaultClient keyVaultClient, String vaultUrl) {
+        super(name, keyVaultClient);
+        this.keyVaultClient = keyVaultClient;
+        this.vaultUrl = vaultUrl;
+
+        listSecretsFromKeyVault();
     }
 
     @Override
     public String[] getPropertyNames() {
-        return new String[0];
+        return this.secretNames.keySet().toArray(new String[this.secretValues.size()]);
     }
 
     @Override
-    public Object getProperty(String s) {
-        return null;
+    public Object getProperty(String name) {

```

Comment:
As public method, not sure null check is need here.

Reply:
Nice catch. We should check it.